tokio-util/src/sync/cancellation_token/implementation.rs

```diff
@@ -0,0 +1,355 @@
+                    // children can no longer be cancelled through us.
+                    // (as we now have neither a parent nor handles)
+                    // Therefore we can disconnect them.
+                    disconnect_children(&mut node);
+                }
+            }
+        });
+    }
+}
+
+/// Cancels a node and its children.
+///
+/// Invariant #1:
+/// It's important that during the entire process of cancellation, there is
+/// never a gap in lock propagation. We need to lock both the parent and the child for the
+/// entire duration of cancellation and disconnection. Those cannot happen separately,
+/// otherwise other calls like [decrease_handle_refcount] could modify the tree structure
+/// while we recurse over it.
+///
+/// We basically have to create a "lock wave" from the parent to children, that nothing else
+/// can bypass. Otherwise, other actions could add/remove children that our cancellation would miss.
+///
+/// Example: Holding a reference of a child does not influence its handle reference count, so
+/// it will happily drop its grandchildren in [decrease_handle_refcount] before we got the chance
+/// to propagate the cancellation to them, if we violate invariant #1.
+/// Unless we keep the child locked during the entire time, from the moment on when we
+/// disconnected it from its parent.
+pub(crate) fn cancel(node: &Arc<TreeNode>) {
+    let mut locked_node = node.inner.lock().unwrap();
+
+    // One by one, adopt grandchildren and then cancel and detach the child
+    while let Some(child) = locked_node.children.pop() {
+        // `node` is already locked, so we can lock the child
+        let mut locked_child = child.inner.lock().unwrap();
+
+        // Detach the child from node
+        // No need to modify node.children, as the child already got removed with `.pop`
+        locked_child.parent = None;
+        locked_child.parent_idx = 0;
+
+        // Cancel or adopt grandchildren
+        while let Some(grandchild) = locked_child.children.pop() {
+            let mut locked_grandchild = grandchild.inner.lock().unwrap();
+
+            // For performance reasons, only adopt grandchildren that have children.
+            // Otherwise, just cancel them right away, no need for another iteration.
+            if locked_grandchild.children.is_empty() {
+                // Detach the grandchild
+                locked_grandchild.parent = None;
+                locked_grandchild.parent_idx = 0;
+
+                // Cancel the grandchild
+                locked_grandchild.is_cancelled = true;
+                drop(locked_grandchild);
+                grandchild.waker.notify_waiters();
+            } else {
+                // Otherwise, adopt grandchild
+                locked_grandchild.parent = locked_child.parent.clone();
+                locked_grandchild.parent_idx = locked_node.children.len();
+                drop(locked_grandchild);
+                locked_node.children.push(grandchild);
```

Comment:
This is the same, but I find it easier to read this way.
```suggestion
                locked_grandchild.parent = node.clone();
                locked_grandchild.parent_idx = locked_node.children.len();
                drop(locked_grandchild);
                locked_node.children.push(grandchild);
```

Reply:
Not possible, is borrowed

Reply:
I would have preferred your version as well. But it's borrowed in the `locked_node`

Reply:
Never mind. My bad.